### PR TITLE
[llvm][CodeGen] Fix failure in window scheduler caused by weak dependencies

### PR DESCRIPTION
**Pull request by huaatian**
This commit addresses an issue where weak dependencies trigger an
assertion in the window scheduler under certain conditions.

### CODE REVIEW

**Comment by huaatian** [comment]
The issue arises because the predecessor of SU(1) is SU(3), which is a weak dependency. However, weak dependencies do not affect the result of list scheduling, so SU(1) is scheduled before SU(3). When calculating the cycle of SU(1), the cycle of SU(3) is not available, leading to the assertion error.
We referred to the swing algorithm, which ignores this kind of weak dependencies, and provided a corresponding solution. It should be noted that list scheduling does not adjust the order of instructions based on weak dependencies ; otherwise, this error would not have occurred since window scheduler use list scheduling result.

Swing scheduler DAG：
image

Window scheduler DAG：
image

List scheduler result：
image

**Comment by huaatian** [comment]
Could you please review this code? Thank you~  @nathanchance @arsenm @dtcxzyw

File: llvm/test/CodeGen/Hexagon/swp-ws-weak-dep.mir [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# REQUIRES: asserts
+# RUN: llc --march=hexagon %s -run-pass=pipeliner -debug-only=pipeliner \
+# RUN: -window-sched=force -o - 2>&1 | FileCheck %s
```
Comment (arsenm) [inline llvm/test/CodeGen/Hexagon/swp-ws-weak-dep.mir:3]: Use -filetype=null since stdout is ignored

**Review by arsenm — APPROVED** [review]

File: llvm/test/CodeGen/Hexagon/swp-ws-weak-dep.mir (hunk 1 above)
Comment (huaatian) [reply llvm/test/CodeGen/Hexagon/swp-ws-weak-dep.mir]: Updated

**Comment by huaatian** [comment]
Currently, this test case fails locally (clang/test/CodeGenCXX/debug-info-ptr-to-ptr.cpp). Due to this commit (https://github.com/llvm/llvm-project/pull/95298), even without my modifications, the test case still fails on my local machine.

**Review by arsenm — APPROVED** [review]

**Pull request merged**